drivers/metal/rendering_device_driver_metal.mm

```diff
@@ -0,0 +1,3883 @@
+				sd.depthStencilPassOperation = STENCIL_OPERATIONS[p_depth_stencil_state.back_op.pass];
+				sd.depthFailureOperation = STENCIL_OPERATIONS[p_depth_stencil_state.back_op.depth_fail];
+				sd.stencilCompareFunction = COMPARE_OPERATORS[p_depth_stencil_state.back_op.compare];
+				sd.readMask = p_depth_stencil_state.back_op.compare_mask;
+				sd.writeMask = p_depth_stencil_state.back_op.write_mask;
+				ds_desc.backFaceStencil = sd;
+			}
+		}
+
+		pipeline->depth_stencil = [device newDepthStencilStateWithDescriptor:ds_desc];
+		ERR_FAIL_NULL_V_MSG(pipeline->depth_stencil, PipelineID(), "Failed to create depth stencil state");
+	} else {
+		// TODO(sgc): FB13671991 raised as Apple docs state calling setDepthStencilState:nil is valid, but currently generates an exception
+		pipeline->depth_stencil = get_resource_cache().get_depth_stencil_state(false, false);
+	}
+
+	// Blend state.
+	{
+		for (uint32_t i = 0; i < p_color_attachments.size(); i++) {
+			if (p_color_attachments[i] == ATTACHMENT_UNUSED) {
+				continue;
+			}
+
+			const PipelineColorBlendState::Attachment &bs = p_blend_state.attachments[i];
+
+			MTLRenderPipelineColorAttachmentDescriptor *ca_desc = desc.colorAttachments[p_color_attachments[i]];
+			ca_desc.blendingEnabled = bs.enable_blend;
+
+			ca_desc.sourceRGBBlendFactor = BLEND_FACTORS[bs.src_color_blend_factor];
+			ca_desc.destinationRGBBlendFactor = BLEND_FACTORS[bs.dst_color_blend_factor];
+			ca_desc.rgbBlendOperation = BLEND_OPERATIONS[bs.color_blend_op];
+
+			ca_desc.sourceAlphaBlendFactor = BLEND_FACTORS[bs.src_alpha_blend_factor];
+			ca_desc.destinationAlphaBlendFactor = BLEND_FACTORS[bs.dst_alpha_blend_factor];
+			ca_desc.alphaBlendOperation = BLEND_OPERATIONS[bs.alpha_blend_op];
+
+			ca_desc.writeMask = MTLColorWriteMaskNone;
+			if (bs.write_r) {
+				ca_desc.writeMask |= MTLColorWriteMaskRed;
+			}
+			if (bs.write_g) {
+				ca_desc.writeMask |= MTLColorWriteMaskGreen;
+			}
+			if (bs.write_b) {
+				ca_desc.writeMask |= MTLColorWriteMaskBlue;
+			}
+			if (bs.write_a) {
+				ca_desc.writeMask |= MTLColorWriteMaskAlpha;
+			}
+		}
+
+		pipeline->raster_state.blend.r = p_blend_state.blend_constant.r;
+		pipeline->raster_state.blend.g = p_blend_state.blend_constant.g;
+		pipeline->raster_state.blend.b = p_blend_state.blend_constant.b;
+		pipeline->raster_state.blend.a = p_blend_state.blend_constant.a;
+	}
+
+	// Dynamic state.
+
+	if (p_dynamic_state.has_flag(DYNAMIC_STATE_DEPTH_BIAS)) {
+		pipeline->raster_state.depth_bias.enabled = true;
+	}
+
+	if (p_dynamic_state.has_flag(DYNAMIC_STATE_BLEND_CONSTANTS)) {
+		pipeline->raster_state.blend.enabled = true;
+	}
+
+	if (p_dynamic_state.has_flag(DYNAMIC_STATE_DEPTH_BOUNDS)) {
+		// TODO(sgc): ??
+	}
+
+	if (p_dynamic_state.has_flag(DYNAMIC_STATE_STENCIL_COMPARE_MASK)) {
+		// TODO(sgc): ??
+	}
+
+	if (p_dynamic_state.has_flag(DYNAMIC_STATE_STENCIL_WRITE_MASK)) {
+		// TODO(sgc): ??
+	}
```

Comment:
These TODOs should likely be tracked in an issue.

Reply:
I don't think they are necessary in Metal – I need to confirm what their equivalents are in Vulkan and make the call. Thanks for the reminder.